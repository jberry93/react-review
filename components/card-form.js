import React, { PureComponent } from "react";

export class CardForm extends PureComponent {
  constructor (props) {
    super(props);
    this.state = {
      card: {
        question: "",
        answer: ""
      },
      hasError: false
    }
  }

  // Currying ftw :D
  handleChange = (propName) => (e) => {
    const { card } = this.state;
    const newCard = {
      ...card,
      [propName]: e.target.value
    };
    this.setState({ card: newCard });
  }

  emptyNewCard = () => {
    this.setState({
      card: {
        question: "",
        answer: ""
      }
    });
  }

  setError = (setting) => {
    this.setState({ hasError: setting });
  }

  submitCard = () => {
    const { card } = this.state;
    const { cards } = this.props;
    if (card.question && card.answer) {
      cards.push(card);
      this.props.updateCards(cards);
      this.emptyNewCard();
      this.setError(false);
    } else {
      this.setError(true);
    }
  }

  render = () => (
    <div className="new-card-form">
      {this.state.hasError &&
        <p className="error-msg">Please give a question and answer</p>
      }
      <div className="form-container">
        <textarea
          placeholder="Front"
          onChange={this.handleChange('question')}
          value={this.state.card.question}
          rows="5"
        ></textarea>
        <textarea
          placeholder="Back"
          onChange={this.handleChange('answer')}
          value={this.state.card.answer}
          rows="5"
        ></textarea>
      </div>
      <div className="action-btns">
        <button className="cancel-btn" onClick={this.props.toggleNewCard}>Cancel</button>
        <button className="create-btn" onClick={this.submitCard}>Create</button>
      </div>
      <style jsx>{`
        .action-btns {
          display: flex;
          flex-direction: row;
          justify-content: center;
        }
        .action-btns > button {
          margin: 0 0.5rem;
        }
        .error-msg {
          color: #F44336;
          margin-bottom: 0;
        }
        .new-card-form {
          display: flex;
          flex-direction: column;
        }
        .form-container {
          display: flex;
          justify-content: space-evenly;
          flex-wrap: wrap;
        }
        .create-btn,
        .cancel-btn {
          width: 6rem;
          margin: 0 auto;
        }
        .create-btn {
          background-color: #00C853;
        }
        .cancel-btn {
          background-color: #D50000;
        }
        textarea {
          border: 1px solid #d3d3d3;
          border-radius: 5px;
          font-family: 'Montserrat', sans-serif;
          height: 10rem;
          margin: 1rem auto;
          outline-width: 0;
          padding: 1rem;
          resize: none;
          transition: 1s;
          width: 25rem;
        }
        textarea:focus {
          background-color: #efefef;
          border: 1px solid #29B6F6;
        }
        @media screen and (max-width: 992px) {
          .form-container {
            justify-content: center;
            flex-direction: column;
            align-items: center;
          }
          .action-btns {
            flex-direction: column;
          }
          .action-btns > button {
            margin: 0.5rem 0;
            width: 100%;
          }
          textarea {
            margin: 0.5rem 0;
            width: calc(100% - 2rem);
          }
        }
      `}</style>
    </div>
  );
}
