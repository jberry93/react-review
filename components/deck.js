import React, { PureComponent } from 'react'

export class Deck extends PureComponent {
  constructor (props) {
    super(props)
    let totalCount = 0
    if (this.props.cards) {
      totalCount = this.props.cards.length
    }
    this.state = {
      toggleSide: true,
      counts: {
        total: totalCount,
        current: 0,
        next: 1,
        prev: 0
      }
    }
  }

  // Need to manually update state when props are updated
  componentWillReceiveProps (props) {
    this.setState({
      counts: {
        total: props.cards.length,
        current: this.state.counts.current,
        next: this.state.counts.next,
        prev: this.state.counts.prev
      }
    })
  }

  switchSides = () => {
    this.setState({ toggleSide: !this.state.toggleSide })
  }

  nextCard = () => {
    this.setState({
      counts: {
        total: this.props.cards.length,
        current: this.state.counts.current < this.state.counts.total - 1 ? this.state.counts.current + 1 : this.state.counts.current,
        next: this.state.counts.next < this.state.counts.total ? this.state.counts.next + 1 : this.state.counts.next,
        prev: this.state.counts.prev < this.state.counts.total - 2 ? this.state.counts.prev + 1 : this.state.counts.prev
      }
    })
  }

  prevCard = () => {
    this.setState({
      counts: {
        total: this.props.cards.length,
        current: this.state.counts.current > 0 ? this.state.counts.current - 1 : 0,
        next: this.state.counts.next > 1 ? this.state.counts.next - 1 : 1,
        prev: this.state.counts.prev > 0 ? this.state.counts.prev - 1 : 0
      }
    })
  }

  render = () => (
    <div className="deck-container">
      <div className="flashcard">
        {this.state.toggleSide ? (
          <strong>{this.props.cards[this.state.counts.current].question}</strong>
        ) : (
          <strong>{this.props.cards[this.state.counts.current].answer}</strong>
        )}
      </div>
      <div className="action-btns">
        <button onClick={this.switchSides}>Flip</button>
        <button onClick={this.prevCard}>Previous</button>
        <button onClick={this.nextCard}>Next</button>
      </div>
      <style jsx>{`
        .action-btns {
          align-items: center;
          display: flex;
          justify-content: space-evenly;
        }
        .action-btns > button {
          margin: 0 3rem;
          width: 6rem;
        }
        .deck-container {
          align-items: center;
          display: flex;
          flex-direction: column;
          justify-content: center;
        }
        .flashcard {
          align-items: center;
          box-shadow: 0 4px 8px 0 rgba(0,0,0,0.2), 0 6px 20px 0 rgba(0,0,0,0.2);
          display: flex;
          flex-wrap: wrap;
          height: 12rem;
          justify-content: center;
          margin: 1rem 0.5rem;
          overflow: auto;
          padding: 1rem;
          width: 36rem;
        }
        @media screen and (max-width: 992px) {
          .deck-container,
          .flashcard,
          .action-btns {
            width: 100%;
          }
          .action-btns > button {
            margin: 0 1rem;
          }
        }
      `}</style>
    </div>
  )
}
