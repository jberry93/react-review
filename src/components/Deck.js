import React, { PureComponent } from 'react'

export class Deck extends PureComponent {
  constructor (props) {
    super(props)
    this.state = {
      toggleSide: true,
      counts: {
        total: this.props.cards.length,
        current: 0,
        next: 1,
        prev: 0
      }
    }
  }

  // Need to manually update state when props are updated
  componentWillReceiveProps (props) {
    this.setState({
      counts: {
        total: props.cards.length,
        current: this.state.counts.current,
        next: this.state.counts.next,
        prev: this.state.counts.prev
      }
    })
  }

  switchSides = () => {
    this.setState({
      toggleSide: !this.state.toggleSide
    })
  }

  nextCard = () => {
    this.setState({
      counts: {
        total: this.props.cards.length,
        current: this.state.counts.current < this.state.counts.total - 1 ? this.state.counts.current + 1 : this.state.counts.current,
        next: this.state.counts.next < this.state.counts.total ? this.state.counts.next + 1 : this.state.counts.next,
        prev: this.state.counts.prev < this.state.counts.total - 2 ? this.state.counts.prev + 1 : this.state.counts.prev
      }
    })
  }

  prevCard = () => {
    this.setState({
      counts: {
        total: this.props.cards.length,
        current: this.state.counts.current > 0 ? this.state.counts.current - 1 : 0,
        next: this.state.counts.next > 1 ? this.state.counts.next - 1 : 1,
        prev: this.state.counts.prev > 0 ? this.state.counts.prev - 1 : 0
      }
    })
  }

  render () {
    return (
      <div>
        {this.state.toggleSide ? (
          <h1>{this.props.cards[this.state.counts.current].question}</h1>
        ) : (
          <h1>{this.props.cards[this.state.counts.current].answer}</h1>
        )}
        <button onClick={this.switchSides}>Flip</button>
        <button onClick={this.prevCard}>Previous</button>
        <button onClick={this.nextCard}>Next</button>
      </div>
    )
  }
}
