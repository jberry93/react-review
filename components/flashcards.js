import React, { PureComponent } from 'react'
import { Deck } from './deck'
import { CardForm } from './card-form'

export class Flashcards extends PureComponent {
  constructor () {
    super()
    this.state = {
      cards: [],
      isNewCard: false
    }
  }

  updateCards = cards => {
    this.setState({ cards })
    this.toggleNewCard()
  }

  toggleNewCard = () => {
    this.setState({ isNewCard: !this.state.isNewCard })
  }

  render = () => (
    <div>
      <h1>React Review</h1>
      <button onClick={this.toggleNewCard}>Create Card</button>
      {this.state.isNewCard &&
        <CardForm
          cards={this.state.cards}
          updateCards={this.updateCards}
        />
      }
      {this.state.cards.length > 0 ? (
        <Deck cards={this.state.cards}/>
      ) : (
        <h1>Sorry there are no cards to show</h1>
      )}
    </div>
  )
}
