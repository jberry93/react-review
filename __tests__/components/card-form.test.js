import React from 'react'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'
import { CardForm } from '../../components/card-form'

describe('CardForm Component', () => {
  it('renders w/out crashing: Enzyme', () => {
    shallow(<CardForm/>)
  })
  it ('renders w/out crashing: Snapshot', () => {
    const component = renderer.create(<CardForm/>)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
