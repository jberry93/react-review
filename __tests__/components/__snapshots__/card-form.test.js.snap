// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardForm Component renders w/out crashing: Snapshot 1`] = `
<div
  className="jsx-2099964395 new-card-form"
>
  <textarea
    className="jsx-2099964395"
    onChange={[Function]}
    placeholder="Question"
    rows="5"
    value=""
  />
  <textarea
    className="jsx-2099964395"
    onChange={[Function]}
    placeholder="Answer"
    rows="5"
    value=""
  />
  <button
    className="jsx-2099964395"
    onClick={[Function]}
  >
    Create
  </button>
</div>
`;
